pageserver/src/tenant/timeline/detach_ancestor.rs

```diff
@@ -352,10 +358,72 @@ pub(super) async fn prepare(
 
     // TODO: copying and lsn prefix copying could be done at the same time with a single fsync after
     let mut new_layers: Vec<Layer> =
-        Vec::with_capacity(straddling_branchpoint.len() + rest_of_historic.len());
+        Vec::with_capacity(straddling_branchpoint.len() + rest_of_historic.len() + 1);
 
     {
-        tracing::debug!(to_rewrite = %straddling_branchpoint.len(), "copying prefix of delta layers");
+        tracing::info!("removing non-inherited keys by writing an image layer at the detach LSN");
+        let io_concurrency = IoConcurrency::spawn_from_conf(
+            detached.conf,
+            detached.gate.enter().map_err(|_| Error::ShuttingDown)?,
+        );
+        let mut reconstruct_state = ValuesReconstructState::new(io_concurrency);
+        // Directly use `get_vectored_impl` to skip the max_vectored_read_key limit check. Note that the keyspace should
+        // not contain too many keys, otherwise this takes a lot of memory.
+        let key_range = Key::sparse_non_inherited_keyspace();
+        // avoid generating a "future layer" which will then be removed
+        let image_lsn = ancestor_lsn;
+        let data = ancestor
+            .get_vectored_impl(
+                KeySpace::single(key_range.clone()),
+                image_lsn,
+                &mut reconstruct_state,
+                ctx,
+            )
+            .await
+            .context("failed to retrieve aux keys")
+            .map_err(|e| Error::launder(e, Error::Prepare))?;
+        if !data.is_empty() {
+            // TODO: is it possible that we can have an image at `image_lsn`? Unlikely because image layers are only generated
```

Comment:
we should probably handle this case, and remove that image layer at that lsn.

an example for having an image would be a prior attempt to detach ancestor where we crashed in between.

Reply:
In such case we should not remove the image layer because it still needs to be used by the parent branch?

I'd rather make this fail for now, though for a small chance it would cause the detach ancestor operation to fail.